sampler.go

```diff
@@ -199,45 +200,74 @@ func (s *rateLimitingSampler) Equal(other Sampler) bool {
 // The probabilisticSampler is given higher priority when tags are emitted, ie. if IsSampled() for both
 // samplers return true, the tags for probabilisticSampler will be used.
 type GuaranteedThroughputProbabilisticSampler struct {
-	probabilisticSampler Sampler
-	lowerBoundSampler    Sampler
-	tags                 []Tag
-	samplingRate         float64
-	lowerBound           float64
+	probabilisticSampler  Sampler
+	lowerBoundSampler     Sampler
+	upperBoundRateLimiter utils.RateLimiter
+	tags                  []Tag
+	samplingRate          float64
+	lowerBound            float64
+	upperBound            float64
 }
 
 // NewGuaranteedThroughputProbabilisticSampler returns a delegating sampler that applies both
 // probabilisticSampler and rateLimitingSampler.
 func NewGuaranteedThroughputProbabilisticSampler(
-	lowerBound, samplingRate float64,
+	lowerBound, samplingRate, upperBound float64,
 ) (*GuaranteedThroughputProbabilisticSampler, error) {
-	tags := []Tag{
-		{key: SamplerTypeTagKey, value: SamplerTypeLowerBound},
-		{key: SamplerParamTagKey, value: samplingRate},
-	}
 	probabilisticSampler, err := NewProbabilisticSampler(samplingRate)
 	if err != nil {
```

Comment:
maybe we should consider defining private constructors for all samplers and using those internally. In the next major we can deprecate the public constructors that don't make sense to be exposed.